### PR TITLE
Split api send tests

### PR DESCRIPTION
**Pull request by jerith**
Almost all of the `test_send` methods in the `api.test` module currently fail due to duplicate send prevention. This branch splits each of those tests into two methods (one for success, one for failure) to avoid sending the same message multiple times and to keep the message state for the two cases separate.

### CODE REVIEW

**Comment by nicpottier** [comment]
Hrmm, these don't actually fail for us. Wonder what is different there. 

How are you running them?

**Comment by jerith** [comment]
I run `python manage.py test` after following http://rapidpro.github.io/rapidpro/docs/development/ and adding the postgis extension to the default database template. I see now that the `settings.py` I'm using overrides `REDIS_DB`, but `clear_cache()` hardcodes `db=10`.

This split is probably worthwhile in any case, though, because the success and error cases are different behaviours that shouldn't really share message state.

**Comment by nicpottier** [comment]
Ok sure.. maybe worth putting the various channel setup bits in setup() for these now that we are repeating those bits on more time?

**Comment by jerith** [comment]
I've factored out the channel setup (and a couple of other things common to all channel type tests) into a new base class and cleaned up a few inconsistencies in the process.

File: temba/api/tests.py [hunk 1]
```diff
@@ -1803,9 +1813,7 @@ def test_send(self):
         # our outgoing sms
         sms = bcast.get_messages()[0]
 
-        try:
-            settings.SEND_MESSAGES = True
-
+        with self.allow_message_sending():
```
Comment (nicpottier) [inline temba/api/tests.py]: Nice, much cleaner.

File: temba/api/tests.py [hunk 2]
```diff
@@ -3166,8 +3183,15 @@ def test_send(self):
...
                 self.assertTrue(msg.sent_on)
 
-                self.clear_cache()
+    def test_send_error(self):
+        # This tests a send error.
+        joe = self.create_contact("Joe", number="+250788383383", twitter="joe1981")
+        bcast = joe.send("Test message", self.admin, trigger_send=False)
+
+        # our outgoing message
+        msg = bcast.get_messages()[0]
```
Comment (nicpottier) [inline temba/api/tests.py]: Seems like these three lines are a good candidate to pull into the base class?

**Comment by nicpottier** [comment]
This looks good, for extra credit would be nice to pull out those repeated lines as well, but I'm fine as is, that isn't your mess. :)

Can you send us a signed contributor agreement?
  https://github.com/rapidpro/rapidpro/blob/master/CONTRIBUTING.md

File: temba/api/tests.py (hunk 2 above)
Comment (jerith) [inline temba/api/tests.py]: I thought about it, but there are lots of differences between the various `self.create_contact()` and `joe.send()` calls. Factoring them out into a method on each test class doesn't really buy very much. For the tests already had them in `.setUp()` I moved them to `.setup_channel()`, but I'm not sure they really belong there. (If they do, the method should probably be renamed.)

Comment (nicpottier) [inline temba/api/tests.py]: Fair enough. Admittedly these are some serious utility tests that were generally written in a rush as we were integrating a new partner, so ya, they could use some more love to normalize it all.

But you are definitely leaving them better than you found them. :+1:

**Comment by nicpottier** [comment]
PS. And if you merge in rapidpro/master should be able to get Travis giving us a build status.

**Comment by nicpottier** [comment]
Closing as it doesn't seem CAA is being signed.

**Pull request closed without merging**